File: programmers/etc/level2/전력망을_둘로_나누기.py
from collections import deque

def solution(n, wires):
    answer = 101
    graph = [[] for _ in range(n+1)]
    for a, b in wires:
        graph[a].append(b)
        graph[b].append(a)
    
    def bfs(node, visited, graph):
        q = deque([node])
        visited[node] = True
        res = 1
        while q:
            cur = q.popleft()
            for nxt in graph[cur]:
                if not visited[nxt]:
                    visited[nxt] = True
                    res += 1
                    q.append(nxt)
        return res
    
    for a, b in wires:
        visited = [False]*(n+1)
        visited[b] = True
        res = bfs(a, visited, graph)
        answer = min(answer, abs(res-(n-res)))
    
    return answer


n = 9
wires = [[1,3],[2,3],[3,4],[4,5],[4,6],[4,7],[7,8],[7,9]]
print(solution(n, wires)) # 3

n = 4
wires = [[1,2],[2,3],[3,4]]
print(solution(n, wires)) # 0

n = 7
wires = [[1,2],[2,7],[3,7],[3,4],[4,5],[6,7]]
print(solution(n, wires)) # 1

File: Java/programmers/etc/level2/전력망을둘로나누기.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 전력망을둘로나누기 {
    public int solution(int n, int[][] wires) {
        int answer = 101;
        List<List<Integer>> graph = new ArrayList<>();
        
        for (int i=0; i<n+1; i++) {
            graph.add(new ArrayList<Integer>());
        }

        for (int[] wire : wires) {
            graph.get(wire[0]).add(wire[1]);
            graph.get(wire[1]).add(wire[0]);
        }

        for (int[] wire : wires) {
            boolean[] visited = new boolean[n+1];
            visited[wire[1]] = true;
            int res = bfs(wire[0], visited, graph);
            answer = Math.min(answer, Math.abs(res-(n-res)));
        }

        return answer;
    }

    private int bfs(int node, boolean[] visited, List<List<Integer>> graph) {
        Queue<Integer> q = new LinkedList<>();
        q.add(node);
        visited[node] = true;
        int res = 1;

        while (!q.isEmpty()) {
            int cur = q.poll();
            for (int nxt : graph.get(cur)) {
                if (!visited[nxt]) {
                    visited[nxt] = true;
                    res++;
                    q.add(nxt);
                }
            }
        }

        return res;
    }
}
